### PR TITLE
Add support for filtering listener per flow node instance

### LINKED ISSUE #23941: Display Listeners for specific flow node instance
Part of #21974 epic.

### Description
For execution listeners, wether the user selected a flow node on the diagram or a flow node instance in the Instance History panel, the list of Listeners displayed would show all the Listeners attached to that flow node. That is because we used only the `flowNodeId` to get the list of Listeners.

Now, we are implementing a new field on the request's body called `flowNodeInstanceId`. The change will be specifically for when the user select a flow node instance, which is now going to show only the Listeners for that specific instance on the Listeners tab, instead of all listeners associated with that flow node.

`flowNodeId` and `flowNodeInstanceId` will be mutually exclusive, meaning whenever we have the most specific of the two (`flowNodeInstanceId`), that's the only one that will be included in the payload and if we don't have it, then only the `flowNodeId` will be included in the request's payload.

### User Story

```Gherkin
Scenario: As an Operate user
    Given I am logged in and in the process instance page
    When I select a flow node in the Instance History panel
    Then the Listeners tab should show only the Listeners related to that flow node instance
```

### Additional information

Is blocked by #23942

### Acceptance Criteria

- [ ] When selecting a flow node on the diagram I am still able to see all listeners related to it on the Listeners tab
- [ ] When selecting a flow node instance in the Instance History panel, I am now able to see only the listeners related to that instance

### Definition of Ready - Checklist

- [x] The issue has a meaningful title, description, and testable acceptance criteria
- [ ] If documentation needs to be updated, an issue is created in the [camunda-docs](https://github.com/camunda/camunda-docs) repo, and the issue is added to our Operate project board.
- [ ] If HELM charts need to be updated, an issue is created in the [camunda-platform-heml](https://github.com/camunda/camunda-platform-helm) repo, and the issue is added to our Operate project board.
- [ ] Potential Security risks have been considered and added to the product-hub issue risk assessment
- [ ] Cross-team dependencies have been considered

Optional:
- [ ] Design input has been collected by the assignee
#### How to test:

1. Find or create a process that has listeners attached to a task, navigate to the process instance page.
2. Using the `Modify instance` feature add a new flow node instance to the task with the listener.
3. When the new instance appears on the Instance History panel, check that:
    a) selecting the flow node in the diagram shows all listeners (for all instances) on the Listeners tab;
    b) selecting a flow node instance in the Instance History panel shows only listeners related to that instance on the 
         Listeners tab.

- Docker image to test: `registry.camunda.cloud/team-operate/camunda-operate:branch-fe-23941-flow-node-instance-listeners`
- Release version: `8.7-alpha2`

### :green_book: Link to the test case 
[Users can view dedicated listeners for each flow node instance](https://camunda.testrail.com/index.php?/cases/view/531015)

### PR DESCRIPTION
**Pull request by vitorwtavares**
## Description

- Adds new (optional) parameter to the listeners request payload called `flowNodeInstanceId` to be able to get a filtered list of only the listeners corresponding to a specific flow node instance
- Makes `flowNodeId` optional in the listeners request payload
- Adapt store to handle both possibilities
- Add new E2E test case to check for difference between getting listeners for a flow node and a flow node instance

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #23941
related to #23942

### CODE REVIEW

File: operate/client/src/App/ProcessInstance/BottomPanel/VariablePanel/index.tsx [hunk 1]
```diff
@@ -37,14 +39,23 @@ const VariablePanel = observer(function VariablePanel() {
...
-    if (flowNodeId) {
+    if (!flowNodeInstanceId && flowNodeId) {
       fetchListeners({
         fetchType: 'initial',
         processInstanceId: processInstanceId,
         payload: {flowNodeId},
       });
+    } else if (
+      flowNodeInstanceId &&
+      !flowNodeSelectionStore.isRootNodeSelected
```
Comment (vitorwtavares) [inline operate/client/src/App/ProcessInstance/BottomPanel/VariablePanel/index.tsx:50]: By default the selection store returns the process instance id as the `flowNodeInstanceId`, this is to avoid a request being made when entering the screen since the root node will never have listeners attached to it.

File: operate/client/src/modules/mocks/mockListenerInstances.ts [hunk 2]
```diff
@@ -171,4 +171,25 @@ const mockListenerInstances: ListenerEntity[] = [
...
+    listenerKey: 'INSTANCE',
+    state: 'ACTIVE',
+    event: 'START',
+    jobType: 'EVENT',
+    time: '2024-01-02 10:53:16',
+    sortValues: ['2'],
+  },
+];
+
+export {mockListenerInstances, mockListenerInstancesShort};
```
Comment (vitorwtavares) [inline operate/client/src/modules/mocks/mockListenerInstances.ts:195]: We will most likely need this mocks again shortly for other listeners-related tasks, so I just decided to leave this in along with the other mock listeners.

File: operate/client/src/modules/stores/processInstanceListeners.test.ts [hunk 3]
```diff
@@ -80,8 +80,6 @@ describe('stores/processInstancesListeners', () => {
   });
 
   it('should fetch next listeners', async () => {
-    mockFetchProcessInstanceListeners().withSuccess(mockListenerInstances);
-
```
Comment (vitorwtavares) [inline operate/client/src/modules/stores/processInstanceListeners.test.ts:84]: This is unrelated to this issue, I just noticed we already do the same thing on the `beforeEach` hook for this test suite, so this is unnecessary. Same situation for lines 147-148.

**Review by pedesen — CHANGES_REQUESTED** [review]
Code looks good, but I found an edge case which is not working anymore:

When there is a multi instance task with listeners, and the user clicks on it from the instance history, the listeners tab is not visible, see attached video: [listeners-tab-2024-10-31_16.25.17.mkv.zip](https://github.com/user-attachments/files/17589684/listeners-tab-2024-10-31_16.25.17.mkv.zip)

File: operate/client/src/modules/stores/processInstanceListeners.ts [hunk 4]
```diff
@@ -175,11 +182,19 @@ class ProcessInstanceListeners {
   fetchPreviousInstances = async () => {
```
[the remark this answers is not in the document]
Comment (vitorwtavares) [reply operate/client/src/modules/stores/processInstanceListeners.ts:182]: I think I was a bit divided on the nomenclature when I was working on the Execution Listeners as to if the items listed on this were the actual listeners or instances of a listener, since the Listener itself would be element attached to the task in the process.

I guess it's just semantics, though, and `Listeners` is more descriptive than `Instances` in this scenario, so I'll commit the change along with the fix for the other comment.

**Comment by vitorwtavares** [comment]
> Code looks good, but I found an edge case which is not working anymore:
> 
> When there is a multi instance task with listeners, and the user clicks on it from the instance history, the listeners tab is not visible, see attached video: [listeners-tab-2024-10-31_16.25.17.mkv.zip](https://github.com/user-attachments/files/17589684/listeners-tab-2024-10-31_16.25.17.mkv.zip)

I believe this is actually a BE issue. I took a look and the `flowNodeInstanceId` in that case is being passed correctly on the request. For some reason the listeners are not bound to the "parent" flow node instance in the multi-instance scenario. 

I initially thought we could just use the `flowNodeId` in this scenario, but after looking into it further the multi-instance flow nodes should behave just like any other flow node **instance** in the Instance History, with the caveat that it should aggregate all listeners for the "child" instances.

I'll create a bugfix issue for that for the backend, in my opinion this should not be a blocker the FE PR.

**Review by pedesen — APPROVED** [review]

**Comment by vitorwtavares** [comment]
@Szik @gastonpillet01 Please test with this image: `registry.camunda.cloud/team-operate/camunda-operate:branch-fe-23941-flow-node-instance-listeners`

I'm trying to re-deploy the preview environment, but I was getting errors and don't have time to investigate right now, so I recommend using the image.

**Review by gastonpillet01 — APPROVED** [review]
Support for filtering listener per flow node instance added.

**Review by Szik — APPROVED** [review]

**Comment by ArezooAzar** [comment]
@vitorwtavares :please review the test case [here](https://camunda.testrail.com/index.php?/cases/view/531015)

**Comment by vitorwtavares** [comment]
> @vitorwtavares :please review the test case [here](https://camunda.testrail.com/index.php?/cases/view/531015)

@ArezooAzar Looks good to me. Please let me know when this the testing is done by approving this PR so I can go ahead and merge it.

**Review by ArezooAzar — APPROVED** [review]
